kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -16,228 +16,71 @@
-    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
-        ctx.fireChannelReadComplete();
-    }
-
-    /**
-     * Calls {@link ChannelHandlerContext#fireUserEventTriggered(Object)} to forward
-     * to the next {@link ChannelInboundHandler} in the {@link ChannelPipeline}.
-     *
-     * Sub-classes may override this method to change behavior.
-     */
-    @Override
-    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-        ctx.fireUserEventTriggered(evt);
-    }
-
-    /**
-     * Calls {@link ChannelHandlerContext#fireChannelWritabilityChanged()} to forward
-     * to the next {@link ChannelInboundHandler} in the {@link ChannelPipeline}.
-     *
-     * Sub-classes may override this method to change behavior.
-     */
-    @Override
-    public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
-        ctx.fireChannelWritabilityChanged();
-    }
-
-    /**
-     * Calls {@link ChannelHandlerContext#fireExceptionCaught(Throwable)} to forward
-     * to the next {@link ChannelHandler} in the {@link ChannelPipeline}.
-     *
-     * Sub-classes may override this method to change behavior.
-     */
-    @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause)
-            throws Exception {
-        ctx.fireExceptionCaught(cause);
+    String filterDescriptor() {
+        return filter.getClass().getSimpleName() + "@" + System.identityHashCode(filter);
     }
 
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
         if (msg instanceof DecodedRequestFrame) {
             DecodedRequestFrame<?> decodedFrame = (DecodedRequestFrame<?>) msg;
-            // Guarding against invoking the filter unexpectely
+            // Guard against invoking the filter unexpectedly
             if (filter.shouldDeserializeRequest(decodedFrame.apiKey(), decodedFrame.apiVersion())) {
-                try (var filterContext = new DefaultFilterContext(ctx, decodedFrame, promise)) {
+                try (var filterContext = new DefaultFilterContext(filter, ctx, decodedFrame, promise, Objects.requireNonNull(outboundCtx))) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug("{}: Dispatching downstream {} request to filter{}: {}",
+                                ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
+                    }
                     filter.onRequest(decodedFrame, filterContext);
                 }
-
             }
             else {
                 ctx.write(msg, promise);
             }
         }
         else {
+            if (!(msg instanceof OpaqueRequestFrame)
+                    && msg != Unpooled.EMPTY_BUFFER) {
+                // Unpooled.EMPTY_BUFFER is used by KafkaProxyFrontendHandler#closeOnFlush
+                // but otherwise we don't expect any other kind of message
```

Comment:
@franz1981 I wasn't sure if the KafkaProxyFrontendHandler#closeOnFlush code really needs to write the empty buffer. Why can't it just call `flush()`?